pkg/kube/controllers/queue.go

```diff
@@ -0,0 +1,173 @@
+func WithRateLimiter(r workqueue.RateLimiter) func(q *Queue) {
+	return func(q *Queue) {
+		q.queue = workqueue.NewRateLimitingQueue(r)
+	}
+}
+
+// WithMaxAttempts allows defining a custom max attempts for the queue. If not set, items will be discarded after 5 attempts
+func WithMaxAttempts(n int) func(q *Queue) {
+	return func(q *Queue) {
+		q.maxAttempts = n
+	}
+}
+
+// WithReconciler defines the to handle items on the queue
+func WithReconciler(f func(name types.NamespacedName) error) func(q *Queue) {
+	return func(q *Queue) {
+		q.workFn = func(key interface{}) error {
+			return f(key.(types.NamespacedName))
+		}
+	}
+}
+
+// NewQueue creates a new queue
+func NewQueue(options ...func(*Queue)) Queue {
+	q := Queue{
+		initialSync: atomic.NewBool(false),
+	}
+	for _, o := range options {
+		o(&q)
+	}
+	if q.name == "" {
+		q.name = "queue"
+	}
+	if q.queue == nil {
+		q.queue = workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter())
+	}
+	if q.maxAttempts == 0 {
+		q.maxAttempts = 5
+	}
+	q.log = log.WithLabels("controller", q.name)
+	return q
+}
+
+// Add an item to the queue.
+func (q Queue) Add(item types.NamespacedName) {
+	q.queue.Add(item)
+}
+
+// AddObject takes an Object and adds the types.NamespacedName associated.
+func (q Queue) AddObject(obj Object) {
+	q.queue.Add(types.NamespacedName{
+		Namespace: obj.GetNamespace(),
+		Name:      obj.GetName(),
+	})
+}
+
+// Run the queue. This is synchronous, so should typically be called in a goroutine.
+func (q Queue) Run(stop <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer q.queue.ShutDown()
+	q.log.Infof("starting")
+	q.queue.Add(defaultSyncSignal)
+	done := make(chan struct{})
+	go func() {
+		// Process updates until we return false, which indicates the queue is terminated
+		for q.processNextItem() {
+		}
+		close(done)
+	}()
+	select {
+	case <-stop:
+	case <-done:
+	}
+	q.log.Infof("stopped")
+}
+
+// syncSignal defines a dummy signal that is enqueued when .Run() is called. This allows us to detect
+// when we have processed all items added to the queue prior to Run().
+type syncSignal struct{}
+
+// defaultSyncSignal is a singleton instanceof syncSignal.
+var defaultSyncSignal = syncSignal{}
+
+// HasSynced returns true if the queue has 'synced'. A synced queue has started running and has
+// processed all events that were added prior to Run() being called Warning: these items will be
+// processed at least once, but may have failed.
+func (q Queue) HasSynced() bool {
+	return q.initialSync.Load()
+}
+
+// processNextItem is the main workFn loop for the queue
+func (q Queue) processNextItem() bool {
+	// Wait until there is a new item in the working queue
+	key, quit := q.queue.Get()
+	if quit {
+		// We are done, signal to exit the queue
+		return false
+	}
+
+	// We got the sync signal. This is not a real event, so we exit early after signaling we are synced
+	if key == defaultSyncSignal {
+		q.log.Debugf("synced")
```

Comment:
Same comment ... starting to feel like we should just provide a name to all of the queues?